modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc

```diff
@@ -0,0 +1,358 @@
+
+* If there is no UNION/UNION ALL separation for the recursive CTE, the query defaults to a normal CTE.
+
+* A syntax error is returned when optional subclauses are used without the RECURSIVE keyword.
+
+* In general, recursion is also limited by:
+** The logic in the recursive statement.
+** The stop in the options argument.
+** Breaching the request timeout, if configured.
+** Breaching the request memory quota, if configured.
+** Exceeding the implicit document limit (10000) when a memory quota is not in use and when no explicit document limit is set in the options.
+** Exceeding the implicit level limit (1000) when the level option is not in use.
+
+[#examples_section]
+== Examples
+
+The following examples follow linear recursion.
+Only one recursive reference is allowed in the FROM clause.
+Self-joins or set-ops are not allowed in the recursive reference.
+
+.Example of simple recursive referencing
+====
+[source,sqlpp]
+----
+WITH RECURSIVE cte AS (
+    SELECT 1 AS r 
+        UNION 
+    SELECT cte.r+1 AS r 
+    FROM cte 
+    WHERE cte.r<4 
+) 
+SELECT cte.r FROM cte;
+----
+
+.Results
+[source,json]
+----
+[
+  {
+    "r": 1
+  },
+  {
+    "r": 2
+  },
+  {
+    "r": 3
+  },
+  {
+    "r": 4
+  }
+]
+----
+====
+
+.Example combining recursive and non-recursive CTEs with the WITH clause
+====
+[source,sqlpp]
+----
+WITH RECURSIVE cte AS (SELECT 1 r) ,
+    rcte AS ( 
+        SELECT cte.r FROM cte 
+            UNION 
+        SELECT rcte.r+2 r FROM rcte WHERE rcte.r<7
+    )
+SELECT * FROM rcte;
+----
+
+.Results
+[source,json]
+----
+[
+  {
+    "rcte": {
+      "r": 1
+    }
+  },
+  {
+    "rcte": {
+      "r": 3
+    }
+  },
+  {
+    "rcte": {
+      "r": 5
+    }
+  },
+  {
+    "rcte": {
+      "r": 7
+    }
+  }
+]
+----
+====
+
+.Examples using the OPTIONS clause and the CYCLE clause to avoid infinite recursion
+====
+.Example resulting in infinite recursion
+[source,sqlpp]
+----
+WITH RECURSIVE similar_items AS (
+  SELECT p.*, 0 as lvl FROM products p WHERE p.id=1
+    UNION
+  SELECT p1.*, similar_items.lvl+1 as lvl FROM products p1,
+    similar_items WHERE p1.id IN similar_items.related_items
+)
+SELECT * FROM similar_items;
+----
+
+.Solution using the OPTIONS clause to add a level limit
+[source,sqlpp]
+----
+WITH RECURSIVE similar_items AS (
+  SELECT p.*, 0 as lvl FROM products p WHERE p.id=1
+    UNION
+  SELECT p1.*, similar_items.lvl+1 as lvl FROM products p1,
+    similar_items WHERE p1.id IN similar_items.related_items
+)
+OPTIONS {"levels:2"}
+SELECT * FROM similar_items;
+---- 
+
+.Solution using the CYCLE clause to track which fields are likely to repeat and cause a cycle
+[source,sqlpp]
+----
+WITH RECURSIVE similar_items AS (
+  SELECT p.*, 0 as lvl FROM products p WHERE p.id=1
+    UNION
+  SELECT p1.*, similar_items.lvl+1 as lvl FROM products p1,
+    similar_items WHERE p1.id IN similar_items.related_items
+)
+CYCLE id RESTRICT
+SELECT * FROM similar_items;
+---- 
+====
```

Comment:
@simon-dew Hey Simon, I added these examples here to show how to use OPTIONS and CYCLE to avoid causing infinite recursion. Let me know what you think! :)